torch/distributed/pipelining/schedules.py

```diff
@@ -684,43 +698,47 @@ def _step_microbatches(
-                        ops.extend(stage.get_bwd_recv_ops(mb_index))
-                else:
-                    raise ValueError(f"Unknown computation type {computation_type}")
+            for prev_rank in all_prev_ranks:
+                prev_rank_ops = self.pipeline_order[prev_rank]
+                prev_rank_action = None
+                if time_step < len(prev_rank_ops):
+                    prev_rank_action = prev_rank_ops[time_step]
+                if prev_rank_action is not None:
+                    computation_type, mb_index, stage_index = prev_rank_action
+                    # Only handle sends for the forward from a previous rank
+                    if computation_type == _ComputationType.FORWARD:
+                        # If not the last stage, then receive fwd activations
+                        if stage_index + 1 in stage_index_to_stage:
+                            # TODO: We are assuming that stage will always receive from stage-1
+                            # however that is not necessarily true of get_fwd_recv_ops
+                            stage = stage_index_to_stage[stage_index + 1]
+                            ops.extend(stage.get_fwd_recv_ops(mb_index))
+                    elif computation_type == _ComputationType.BACKWARD:
+                        # Previous rank doing backward has no influence for the current rank forward recv
+                        pass
+                    else:
+                        raise ValueError(f"Unknown computation type {computation_type}")
+
+            for next_rank in all_next_ranks:
+                next_rank_ops = self.pipeline_order[next_rank]
+                next_rank_action = None
+                if time_step < len(next_rank_ops):
+                    next_rank_action = next_rank_ops[time_step]
+                if next_rank_action is not None:
+                    computation_type, mb_index, stage_index = next_rank_action
+                    # Only handle receives for the backwards from a next rank
+                    if computation_type == _ComputationType.FORWARD:
+                        # Next rank doing forward has no influence for the current rank backward recv
+                        pass
+                    elif computation_type == _ComputationType.BACKWARD:
+                        # If not the first stage, then receive bwd gradients
+                        if stage_index - 1 in stage_index_to_stage:
+                            # TODO: We are assuming that stage will always receive from stage+1
+                            # however that is not necessarily true of get_bwd_recv_ops
+                            stage = stage_index_to_stage[stage_index - 1]
+                            ops.extend(stage.get_bwd_recv_ops(mb_index))
+                    else:
+                        raise ValueError(f"Unknown computation type {computation_type}")
 
             # do the communication
             if ops:
```

Comment:
something unrelated to the current PR but might be an issue with the runtime approach we have taken-

we batch all the comm ops for the current timestep, and then issue them after the current action.  Then we come back to do another action.  It looks like we would not have any opportunity to overlap comms.

Specifically, we seem to have this pattern

Action()
Comms().wait()
Action()
comms().wait()

Which IIUC would result in a cuda timeline like 
[Action][Comms][Action][Comms]...

What if we want to do some of the comms so they overlap with some of the actions?